src/FSharp.Data.GraphQL.Server/Values.fs

```diff
@@ -89,10 +89,35 @@ let rec internal compileByType (inputObjectPath: FieldPath) (inputSource : Input
 
     | InputObject objDef ->
         let objtype = objDef.Type
-        let ctor = ReflectionHelper.matchConstructor objtype (objDef.Fields |> Array.map (fun x -> x.Name))
+        let (constructor : obj[] -> obj), (parameterInfos : Reflection.ParameterInfo[]) =
+            if typeof<IDictionary<string,obj>>.IsAssignableFrom(objtype) then
```

Comment:
I would also check for `IReadOnlyDictionary<string, obj>` and `obj` and create `ImmutableDictionary<string, obj>` in that case